File: compiler/bench/levenshtein.fs
\ expected: 60000
\ Edit distance, 10K pairs of short strings (length 8)
\ Checksum: sum of all distances

create s1 16 allot
create s2 16 allot
create dp 81 cells allot

: min2 ( a b -- min ) 2dup > if swap then drop ;
: min3 ( a b c -- min ) min2 min2 ;
: dp@ ( i j -- val ) 9 * + cells dp + @ ;
: dp! ( val i j -- ) 9 * + cells dp + ! ;

: fill-s1 ( n -- ) 8 0 do i n + 26 mod 97 + s1 i + c! loop ;
: fill-s2 ( n -- ) 8 0 do i n 3 + + 26 mod 97 + s2 i + c! loop ;

: levenshtein ( -- dist )
  9 0 do i 0 i dp! loop
  9 0 do 0 i 0 dp! loop
  8 0 do
    8 0 do
      s1 j + c@ s2 i + c@ = if
        j i dp@ j 1+ i 1+ dp!
      else
        j i dp@ 1+
        j 1+ i dp@ 1+
        j i 1+ dp@ 1+
        min3 j 1+ i 1+ dp!
      then
    loop
  loop
  8 8 dp@ ;

: bench-lev ( -- sum )
  0
  10000 0 do
    i fill-s1
    i fill-s2
    levenshtein +
  loop ;

: main bench-lev . cr ;

File: compiler/bench/levenshtein.c
// expected: 60000
// Edit distance, 10K pairs of short strings (length 8)
// Checksum: sum of all distances

#include <stdio.h>

char s1[16];
char s2[16];
int dp[9][9];

int min2(int a, int b) { return a < b ? a : b; }
int min3(int a, int b, int c) { return min2(min2(a, b), c); }

void fill_s1(int n) {
    for (int i = 0; i < 8; i++) {
        s1[i] = 'a' + ((i + n) % 26);
    }
}

void fill_s2(int n) {
    for (int i = 0; i < 8; i++) {
        s2[i] = 'a' + ((i + n + 3) % 26);
    }
}

int levenshtein(void) {
    for (int i = 0; i <= 8; i++) {
        dp[i][0] = i;
        dp[0][i] = i;
    }
    for (int j = 0; j < 8; j++) {
        for (int i = 0; i < 8; i++) {
            if (s1[j] == s2[i]) {
                dp[j+1][i+1] = dp[j][i];
            } else {
                dp[j+1][i+1] = min3(dp[j][i] + 1, dp[j+1][i] + 1, dp[j][i+1] + 1);
            }
        }
    }
    return dp[8][8];
}

int main(void) {
    long sum = 0;
    for (int i = 0; i < 10000; i++) {
        fill_s1(i);
        fill_s2(i);
        sum += levenshtein();
    }
    printf("%ld\n", sum);
    return 0;
}
